### PR TITLE
trying to improve bill listing

### PR DESCRIPTION
**Pull request by devfaz**
Fixes issue => https://github.com/orgs/firefly-iii/discussions/7655#discussioncomment-6789382

Changes in this pull request:

- bills are no longer listed "in the wrong month" if bill-date is start/end of month
- 28th of Feb no longer destroys calculation of next-billing-date

I tested a lot and it looks afaik good for me. Maybe you can re-use some of these changes?

@JC5

### CODE REVIEW

**Comment by devfaz** [comment]
Have to rework a bit.. still issues :|

**Comment by devfaz** [comment]
a small fix which seems to work.

File: app/Http/Controllers/HomeController.php [hunk 1]
```diff
@@ -69,14 +69,14 @@ public function dateRange(Request $request): JsonResponse
     {
         try {
             $stringStart = e((string)$request->get('start'));
-            $start       = Carbon::createFromFormat('Y-m-d', $stringStart);
+            $start       = Carbon::createFromFormat('Y-m-d H:i:s', "$stringStart 00:00:00");
```
Comment (JC5) [inline app/Http/Controllers/HomeController.php:72]: You can also force the start of the day by using `Carbon::createFromFormat('!Y-m-d', $stringStart);`

File: app/Transformers/BillTransformer.php [hunk 2]
```diff
@@ -159,47 +159,48 @@ protected function paidData(Bill $bill): array
                 'next_expected_match' => null,
             ];
         }
-        // 2023-07-1 sub one day from the start date to fix a possible bug (see #7704)
-        // 2023-07-18 this particular date is used to search for the last paid date.
-        // 2023-07-18 the cloned $searchDate is used to grab the correct transactions.
```
Comment (JC5) [inline app/Transformers/BillTransformer.php:164]: Please do not remove all of these comments I put them there for a reason.

File: app/Transformers/BillTransformer.php [hunk 3]
```diff
@@ -159,47 +159,48 @@ protected function paidData(Bill $bill): array
...
         /** @var Carbon $start */
-        $start       = clone $this->parameters->get('start');
-        $searchStart = clone $start;
-        $start->subDay();
```
Comment (JC5) [inline app/Transformers/BillTransformer.php:168]: Why did you remove this line in relation to issue #7704?

File: app/Transformers/BillTransformer.php [hunk 4]
```diff
@@ -159,47 +159,48 @@ protected function paidData(Bill $bill): array
...
         $lastPaidDate = $this->lastPaidDate($set, $start);
-        //Log::debug(sprintf('Result of lastPaidDate is %s', $lastPaidDate->format('Y-m-d')));
+        //Log::debug(sprintf('Result of lastPaidDate is %s', $lastPaidDate->format('Y-m-d H:i')));
 
         /*
          * The next expected match (nextMatch) is, initially, the bill's date.
          */
         $nextMatch = clone $bill->date;
-        //Log::debug(sprintf('Next match is %s (bill->date)', $nextMatch->format('Y-m-d')));
+        $originalNextMatch = clone $bill->date;
```
Comment (JC5) [inline app/Transformers/BillTransformer.php:182]: What's the point of this extra variable?

File: app/Transformers/BillTransformer.php [hunk 5]
```diff
@@ -263,11 +264,14 @@ protected function payDates(Bill $bill): array
         }
         $set          = new Collection();
         $currentStart = clone $this->parameters->get('start');
-        // 2023-06-23 subDay to fix 7655
-        $currentStart->subDay();
```
Comment (JC5) [inline app/Transformers/BillTransformer.php:267]: Why is this removed?

**Comment by JC5** [comment]
Sorry but this feels messy, I can't merge this. Have you seen my fix in the develop branch? What is left of the original issue? Can you provide me with a test case that I can replicate?

File: app/Transformers/BillTransformer.php (hunk 3 above)
Comment (devfaz) [reply app/Transformers/BillTransformer.php:168]: I will spawn a instance to reproduce the issue mentioned in #7704.

File: app/Transformers/BillTransformer.php (hunk 5 above)
Comment (devfaz) [reply app/Transformers/BillTransformer.php:267]: well, the linked issue is the main reason for this PR and the changes done no longer have a requirement for this "->subDay()".

**Comment by devfaz** [comment]
> Sorry but this feels messy, I can't merge this. Have you seen my fix in the develop branch? What is left of the original issue? Can you provide me with a test case that I can replicate?

Np, my feeling is that these subDay() here, subday() there.. working with start/end-Date which contain time is not the cleanest way.

**Comment by devfaz** [comment]
testing current develop image:

* 28th of month issue is still there (the bill is a monthly bill for 30th of each month)

```
Next expected match | October 28th, 2023
```

trying to reproduce the bill-view issue, too.

**Comment by JC5** [comment]
I tried something else entirely. Checkout this commit:

https://github.com/firefly-iii/firefly-iii/commit/d46041f22e94a67be438c28a61dc2e23ab38eb39#diff-b6201c4e79012f7eaef0e2793e4c05dcf5fc61649a776ef423da4e0c8a125cd9L302

**Comment by devfaz** [comment]
Checked the latest develop-image. Its "better", but not perfect. Its still - sometimes - displays 28th of Oct. Reproduced it on demo: https://demo.firefly-iii.org/bills/show/10 

Seems to be correct (30th of month) until you link the first payment.

**Comment by JC5** [comment]
Nice find! Perhaps thats a different routine? It shouldn't be tho 😬

**Comment by JC5** [comment]
Nice find, I applied the same logic. Does it work for you as well?

**Comment by devfaz** [comment]
Well, Im really sorry to say that :) -  but its "better" but there is still an issue

https://demo.firefly-iii.org/bills/show/10

* Bill monthly for 15th
* next calulcation breaks as soon as you link a payment with an later date (f.e. 16th Sep. 2023)
* expected-next-date: 15th Oct. 2023
* current-next-date returned: 15th Sep 2023

same issue for monthly bills starting at 1st of Month (https://demo.firefly-iii.org/bills/show/11)
* payment was linked to 2nd Sep 2023
* expected-next-date: 1 Oct 2023
* current-next-date returned: 1st Sep 2023.

**Comment by JC5** [comment]
Hah, no worries! It's on the list!

**Comment by JC5** [comment]
Same logic, another try.

**Comment by devfaz** [comment]
Looks good now. At least I couldnt find anything unusual. Thx!

**Comment by JC5** [comment]
Great! All credit to you!

**Comment by devfaz** [comment]
@JC5 sorry to re-open this. I found another issue - its nothing which affects me, but maybe it should be fixed.

* create a "weekly" bill, set skip to "16"
* expected: the week is expected every 16 weeks
* result: the week is expected every week.

https://demo.firefly-iii.org/bills/show/10

**Comment by JC5** [comment]
Nice, I see it! Weird. I would expect one routine to catch them all. I'll check out and see what I did wrong.
